### PR TITLE
Add multi arm causal forest

### PR DESCRIPTION
**Pull request by erikcs**
Add `multi_arm_causal_forest(X, Y, W)` for causal forest estimation with multiple categorical treatments (`W` is a factor vector). It is integrated with the summary function `average_treatment_effect` (and `get_scores` - this warns of low/high propensities [here](https://github.com/erikcs/grf/blob/macf/r-package/grf/R/get_scores.R#L244)).

The components(Relabeling/Splitting/Prediction) are:

* [MultiCausalRelabeling](https://github.com/erikcs/grf/blob/macf/core/src/relabeling/MultiCausalRelabelingStrategy.cpp)

* [MultiRegressionSplitting](https://github.com/erikcs/grf/blob/macf/core/src/splitting/MultiRegressionSplittingRule.cpp)
  The original causal forest uses the `InstrumentalSplittingStrategy` which is identical to standard regression splitting, but with more constraints on the children, we avoid that added complexity in this first round, this can in some way be mimicked with a larger `min.node.size`.

* [MultiCausalPrediction](https://github.com/erikcs/grf/blob/macf/core/src/prediction/MultiCausalPredictionStrategy.cpp)
     Prediction and variance estimates.

(these 3 C++ components will be the same for the next "cape_forest" in #733)

### Details

1. The future performance tweaks are the same as mentioned in a follow-up to #742. The expressions for Relabeling and Prediction are equation (20) and (19) from the [grf paper](https://arxiv.org/pdf/1610.01271v4.pdf).
2. Tuning was tried out in commit 3a71c77 and added zero value, reverted it and suggest leaving it out.

### Tests
* [MultiCausalPredictionStrategyTest.cpp](https://github.com/erikcs/grf/blob/macf/core/test/prediction/MultiCausalPredictionStrategyTest.cpp) Sanity check that single treatment predictions are the same as causal forest

* [MultiCausalRelabelingStrategyTest.cpp](https://github.com/erikcs/grf/blob/macf/core/test/relabeling/MultiCausalRelabelingStrategyTest.cpp) Sanity check that the influence function is correct

* [CharacterizationTest](https://github.com/erikcs/grf/blob/macf/core/test/forest/ForestCharacterizationTest.cpp#L288) locks in behavior

* [test_multi_arm_causal_forest.R](https://github.com/erikcs/grf/blob/macf/r-package/grf/tests/testthat/test_multi_arm_causal_forest.R) Statistical behavior

### Checks

* A [check](https://gist.github.com/erikcs/af50a89669439b27323b680ee6947b13) of the regret simulations from policytree
* MSE [check](https://gist.github.com/erikcs/da2e0e6cdd3d78c6b2329ec2689423b9) compared with causal forest
* An [ATE](https://gist.github.com/erikcs/2ce50d10910c5659a6f9b0230db75ada) check for multiple treatment arms

### Follow-ups

* Make sure variance calculations with sample.weights are correct

* [this minor thing](https://github.com/erikcs/grf/blob/macf/core/test/prediction/MultiCausalPredictionStrategyTest.cpp#L129) 

(To beta-test this feature without git cloning: `devtools::install_github("erikcs/grf", subdir = "r-package/grf", ref = "macf")`)

### CODE REVIEW

**Review by halflearned — COMMENTED** [review]
@erikcs, thanks a lot for this excellent PR. 

I'll let others comment on the specific code changes, but I've been using `multi_action_causal_forest` for several weeks now and thus far the code runs without any major issues, except for very minor ones that @erikcs and I have Slacked about and resolved already. 

In terms of computational performance, the `multi_action_causal_forest` runs several times faster than, e.g., `policytree::multi_causal_forest` (of course). I haven't done extensive tests regarding statistical performance (let me know if that'd be helpful!) but anecdotally I don't see any issues.

**Review by jtibshirani — APPROVED** [review]
I didn't review in great detail, but the PR looks good to me overall -- nice structure and well-tested.

Thinking about the statistical checks you ran, do we want to add anything to 'benchmarks' (larger tests that which we run on every release) ?

File: r-package/grf/bindings/RcppUtilities.cpp [hunk 1]
```diff
@@ -178,7 +178,7 @@ Rcpp::NumericMatrix RcppUtilities::create_error_matrix(const std::vector<Predict
     return Rcpp::NumericMatrix(0);
   }
 
-  size_t prediction_length = first_prediction.size();
+  size_t prediction_length = 1;
```
[the remark this answers is not in the document]
Comment (erikcs) [reply r-package/grf/bindings/RcppUtilities.cpp:181]: Yes, these are for collecting the two scalar entries from `predict::compute_error`:  `raw error` and `debiased error`: these are not vectors, so I changed it to 1

[the remark this answers is not in the document]
Comment (erikcs) [reply r-package/grf/bindings/RcppUtilities.cpp:181]: https://github.com/grf-labs/grf/blob/master/core/src/prediction/OptimizedPredictionStrategy.h#L104

**Comment by erikcs** [comment]
Thanks @jtibshirani, yes maybe some of these longer running checks can be added to the benchmarks file

File: r-package/grf/bindings/RcppUtilities.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jtibshirani) [reply r-package/grf/bindings/RcppUtilities.cpp:181]: Got it, maybe we could avoid using the word 'prediction' here then, as it's confusing and hard-code the 1 ?

[the remark this answers is not in the document]
Comment (erikcs) [reply r-package/grf/bindings/RcppUtilities.cpp]: Yes, makes perfect sense, it is updated now

**Pull request merged**